database/interface.py

```diff
@@ -142,13 +148,102 @@ def update_harvest_source(self, source_id, updates):
             self.db.rollback()
             return None
 
+    def clear_harvest_source(self, source_id):
+        """
+        Clear all datasets related to a harvest source in CKAN, and clean up the
+        harvest_record and harvest_record_error tables.
+        :param source_id: ID of the harvest source to clear
+        """
+
+        def _clear_harvest_records():
+                self.db.query(HarvestRecordError).filter(
+                    HarvestRecordError.harvest_record_id.in_(
+                        self.db.query(HarvestRecord.id).filter_by(harvest_source_id=source_id)
+                    )
+                ).delete(synchronize_session=False)
+                self.db.query(HarvestRecord).filter_by(harvest_source_id=source_id).delete()
+                self.db.commit()
+
+        source = self.db.get(HarvestSource, source_id)
+        if source is None:
+            return "Harvest source not found"
+
+        organization_id = source.organization_id
+
+        records = (
+            self.db.query(HarvestRecord)
+            .filter_by(harvest_source_id=source_id).all()
+        )
+        if not records:
+            return "Harvest source has no records to clear."
+
+        ckan_ids = [record.ckan_id for record in records if record.ckan_id is not None]
```

Comment:
what if  len(ckan_ids) < len(records)? is there an ongoing harvesting activity for this source?

Reply:
when len(ckan_ids) < len(records), one situation is maybe there is harvesting activity at same time, the other situation is when calling package_create api, if there is an error, there is no ckan_id for certain records.

Reply:
So len(ckan_ids) + len(error records) = len(records). We need verify this before sending the list of ckan_ids to ckan to clear. Imagine there is an ongoing harvesting activity and some ckan_id is added after this line, ckan can clear all records according to your id list but the newly added ckan_id will be left untouched and will affect action.package_search result.

Reply:
we may add a check to make sure the number match before clear. Or check this page for any in_progress job etc.

Reply:
added the check for the two situations : len(ckan_ids) + len(error records) = len(records) and check no job in_progress